fix: NetworkSceneManager clearing scene placed NetworkObjects list when ClientSynchronizationMode is additive

NoelStephensUnity opened the pull request:
When the `NetworkSceneManager.ClientSynchronizationMode` is set to `LoadSceneMode.Additive` and the client is being synchronized, `NetworkSceneManager` should not clear its scene placed objects list as there might already be in-scene `NetworkObject`s populated. This can happen if the default active scene is already loaded on the host and client(s) side and before one or more client(s) connect the active scene gets changed on the server side.
Pertains to #2377

[MTT-3363](https://jira.unity3d.com/browse/MTT-3363)
[MTT-5356](https://jira.unity3d.com/browse/MTT-5356)

This also resolves issues related to seamless scene transitions and scene streaming related issues when the `NetworkSceneManager.ClientSynchronizationMode` is set to `LoadSceneMode.Additive`.

This also includes added `NetworkObject` scene migration synchronization capabilities that will synchronize clients when a `NetworkObject` is migrated to a different scene as well as synchronizes late joining clients.  
This also includes the ability to synchronize clients when the currently active scene is changed on the server side along with the ability to set a `NetworkObject` to auto-synchronize itself with the change (i.e. automatically migrates to the new active scene).

[Internal Companion Project](https://github.cds.internal.unity3d.com/noel-stephens/Additive-ClientSynchronizationMode): This project demonstrates the various fixes and additions related to this PR.


## Changelog
### Added
- Added: `NetworkSceneManager.ActiveSceneSynchronizationEnabled` property, disabled by default, that enables client synchronization of server-side active scene changes.
- Added: `NetworkObject.ActiveSceneSynchronization`, disabled by default, that will automatically migrate a `NetworkObject` to a newly assigned active scene.
- Added: `NetworkObject.SceneMigrationSynchronization`, enabled by default, that will synchronize client(s) when a `NetworkObject` is migrated into a new scene on the server side via `SceneManager.MoveGameObjectToScene`.

### Changed
- Updated: `NetworkSceneManager` to migrate dynamically spawned `NetworkObject`s with `DestroyWithScene` set to false into the active scene if their current scene is unloaded.
- Updated: the server to synchronize its local `NetworkSceneManager.ClientSynchronizationMode` during the initial client synchronization.

### Fixed:
- Fixed: issue when the `NetworkSceneManager.ClientSynchronizationMode` is `LoadSceneMode.Additive` and the server changes the currently active scene prior to a client connecting then upon a client connecting and being synchronized the NetworkSceneManager would clear its internal ScenePlacedObjects list that could already be populated.
- Fixed: issue where a client would load duplicate scenes of already preloaded scenes during the initial client synchronization and `NetworkSceneManager.ClientSynchronizationMode` was set to `LoadSceneMode.Additive`.

## Testing and Documentation
- Includes integration test additions and updates:
  - ClientSynchronizationModeTests (added for additive client synchronization mode tests)
  - NetworkObjectSceneMigrationTests (added for `NetworkObject` scene migration tests)
  - ParentDynamicUnderInScenePlacedHelper (updated to test Client Synch Additive + Parenting)
  - ParentingInSceneObjectsTests (updated to test Client Synch Additive + Parenting In-Scene Placed)

NoelStephensUnity committed:
fix

When the ClientSynchronizationMode is additive and the client is being synchronized, we need to not clear the ScenePlacedObjects list as there might already be ScenePlacedObjects populated.

NoelStephensUnity committed:
update

Updating changelog entry with fix.

NoelStephensUnity committed:
fix

A more robust fix for running the server-side NetworkSceneManager in client synchronization mode additive.

NoelStephensUnity committed:
test

Updating existing integration tests and the additive scene loading test to be able to manually test.

SamuelBellomo commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -21,6 +21,7 @@ Additional documentation and release notes are available at [Multiplayer Documen
 - Network prefabs are now stored in a ScriptableObject that can be shared between NetworkManagers, and have been exposed for public access. By default, a Default Prefabs List is created that contains all NetworkObject prefabs in the project, and new NetworkManagers will default to using that unless that option is turned off in the Netcode for GameObjects settings. Existing NetworkManagers will ma ...
 
 ### Fixed
+- Fixed issue when the ClientSynchronizationMode is additive and the server changes the currently active scene prior to a client connecting then upon a client connecting and being synchronized the NetworkSceneManager would clear its internal ScenePlacedObjects list that could already be populated. (#2383)
```
```suggestion
- Fixed issue when the ClientSynchronizationMode is additive and the server changes the currently active scene prior to a client connecting then upon a client connecting and being synchronized the NetworkSceneManager would clear its internal ScenePlacedObjects list that could already be populated. (#2383)
    - This removes the need to resync the whole scene when players disconnect and reconnect while still being in the same scene.
```
Just so users know the benefits this unlocks for them (they don't know the internal workings of NGO)

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/CHANGELOG.md

Co-authored-by: Sam Bellomo <71790295+SamuelBellomo@users.noreply.github.com>

NoelStephensUnity committed:
update

Adding a piece to the puzzle where a client has scenes that don't need to be loaded or is disconnected, scenes are still loaded, and some short period later reconnects but certain scenes on the server-side are no longer loaded... under this scenario the client side will unload those scenes.

Still figuring out how to best distinguish between scenes you want to keep loaded but don't get synchronized and those that do get synchronized and you do want to unload.

NoelStephensUnity committed:
Merge branch 'develop' into fix/clientsynchronizationmode-failure-activescene-changes

NoelStephensUnity committed:
Merge branch 'develop' into fix/clientsynchronizationmode-failure-activescene-changes

NoelStephensUnity committed:
update

Auto scene migration for don't destroy with scene (migrates to the currently active scene when the scene is unloaded) and added auto active scene synchronization where, when set,  they automatically migrate into the new active scene.

Includes auto-scene synchronization for late joining clients and dynamically spawned NetworkObjects.

NoelStephensUnity committed:
Merge branch 'develop' into fix/clientsynchronizationmode-failure-activescene-changes

NoelStephensUnity committed:
update

some minor adjustments due to changes.
Removed the scene migration in spawn as it ended up not really making sense to do it during spawn (i.e. synchronization already handles this and if active scene synchronization is enabled then connected clients will spawn in the proper scene already).
Added some missed asserts on time out to the parent dynamic under inscene placed test and removed a section of initialization code no longer needed.

NoelStephensUnity committed:
update

Removing commented out code.

NoelStephensUnity committed:
style

adding a comment about why we are now synchronizing DestroyWithScene

NoelStephensUnity committed:
revert

Reverting the change with DestroyWithScene (breaking API change).
Still will get synchronized, but users will have to set this themselves.

NoelStephensUnity committed:
update

Removed commented code.
Fixed issue with migrating into the appropriate scene.

NoelStephensUnity committed:
update

After some deliberation over whether to include this update, I decided it would make more sense to include full NetworkObject scene migration synchronization.  This includes the additional changes required to assure that clients are synchronized when a dynamically spawned NetworkObject is moved into a new scene.

NoelStephensUnity committed:
test

The additional integration tests for this PR. (WIP)
Includes 3 additional empty scenes to verify object scene migration.
Also a minor style issue addressed in NetworkSceneManager.

NoelStephensUnity committed:
update

updating the changelog

NoelStephensUnity committed:
update

Adjusting a mix-up in the added section.

NoelStephensUnity committed:
update

Fixing isolation test failure on change log.

NoelStephensUnity committed:
test and fix

Adding tests for NetworkObject scene migration.
Updating scene manager handler to include moving objects from a scene to the DDOL (for unloading a scene only).
Fixing minor issue where AddSceneObject was always passing false for the destroy with scene parameter.

NoelStephensUnity committed:
update and test

Added last test, ClientSynchronizationModeTests, for this PR that validates preloaded scenes will be used and in-scene placed NetworkObjects will be synchronized even when the active scene changes which could change the order in which scenes are loaded and synchronized.

This includes minor adjustments to one of the existing test.
This includes the additional INetworkSceneManager.SetClientSynchronizationMode method that now builds a list of preloaded scenes if the client synchronization mode is additive.
Making the parent dynamic under inscene placed run in client synchronization additive mode again.

NoelStephensUnity committed:
style

removing commented out code.

NoelStephensUnity committed:
fix

Found issue where NetworkSceneManager wasn't unsubscribing from active scene changed notifications.

NoelStephensUnity committed:
Merge branch 'develop' into fix/clientsynchronizationmode-failure-activescene-changes

NoelStephensUnity commented on testproject/Assets/Tests/Runtime/NetworkSceneManager/NetworkObjectSceneMigrationTests.cs:
```diff
@@ -0,0 +1,510 @@
+        {
+            // Disable resynchronization for this test to avoid issues with trying
+            // to synchronize them.
+            NetworkSceneManager.DisableReSynchronization = true;
```
This is integration test specific and is not needed under normal operating conditions (i.e. runtime or in editor).

NoelStephensUnity commented on testproject/Assets/Tests/Runtime/NetworkSceneManager/NetworkObjectSceneMigrationTests.cs:
```diff
@@ -0,0 +1,510 @@
+            // Now change the active scene
+            SceneManager.SetActiveScene(m_ScenesLoaded[1]);
+            // We have to do this
+            Object.DontDestroyOnLoad(m_TestPrefabAutoSynchActiveScene);
```
Another integration test thing where we don't want to destroy the temporary prefab instance during an integration test.

samlucas-unity commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkObject.cs:
```diff
@@ -1118,6 +1172,18 @@ public bool WorldPositionStays
+            /// </summary>
+            public bool DestroyWithScene
+            {
+                get => ByteUtility.GetBit(m_BitField, 6);
```
The 'magic' bitField bit indexes numbers should really be consts or defines somewhere, I can see its how its done above though 😀 .  Might be worth another PR after this one to just tidy that up or pop a task somewhere. Not needed for this PR though.

NoelStephensUnity replied:
Yeah, there is history behind the entire `SceneObject` structure where it originally was to handle the initial serialization and synchronization of in-scene placed `NetworkObject`s.  In all reality, this area of code has been slowly improving over time (much better than when it was first implemented).
But we definitely should have another pass (like we are doing with the systems) at classes (i.e. NetworkObject, NetworkBehaviour, etc) where we make adjustments that don't impact users but can help to shave off processing cycles.
_For the most part, the ByteUtility does aggressive inlining so the bit shifting ends up not being that big of a blip on the over-all processing cost radar relative to everything else._

samlucas-unity replied:
I was thinking more for readability something like this.
private const int k_DestroyWithSceneBitIndex = 6;

get => ByteUtility.GetBit(m_BitField, k_DestroyWithSceneBitIndex);

NoelStephensUnity replied:
Yeah... 😸 (_totally missed that point_)
I totally agree... that is a good idea. 👍

samlucas-unity commented on com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs:
```diff
@@ -385,6 +392,81 @@ public AsyncOperation UnloadSceneAsync(Scene scene, SceneEventProgress sceneEven
+            {
+                ServerSceneHandleToClientSceneHandle.Remove(serverHandle);
+            }
+            else
```
Not sure what the process on errors/asserts is but it might be worth making this fall through and then raise asserts or errors or warnings when ClientSceneHandleToServerSceneHandle and HandleToScene contain members not added via UpdateServerClientSceneHandle.  I'm assuming here that these Dictionaries can only be modified by these two functions.  Actually if thats the case then it might be better to group the containers into a class/struct, have the add/remove function and then just expose the accessors, that way you can be sure no one is going to directly modify the dictionaries in the future (makes the grouping of these containers explicit to anyone working on this code in the future, rather than implicit via these two functions)

NoelStephensUnity replied:
That is a really good point.  Let me look at this again, but I think your suggestion is a good idea.

NoelStephensUnity replied:
This area has several tendrils in various areas. I like the idea, but I am going to resist making the change in this PR as it should really be done as a separate PR.

NoelStephensUnity committed:
Merge branch 'develop' into fix/clientsynchronizationmode-failure-activescene-changes

NoelStephensUnity committed:
changelog fix

Migrating last PRs changelog into the Unreleased section.

NoelStephensUnity committed:
update

Removed the HandleToScene dictionary. It was duplicating ScenesLoaded.
Replaced all places that used HandleToScene with ScenesLoaded.

NoelStephensUnity committed:
Merge branch 'develop' into fix/clientsynchronizationmode-failure-activescene-changes

NoelStephensUnity committed:
Merge branch 'develop' into fix/clientsynchronizationmode-failure-activescene-changes

NoelStephensUnity committed:
update and fix

Fixing an issue when ClientSynchronizationMode is LoadSceneMode.Single and the scene to be loaded is the primary scene and is already loaded on the client side. Caught this issue running the manual test: PreserveNetworkObjectsOnShutdown.

Migrating the "shouldPassThrough" logic to the handler as this behavior/logic is slightly different when integration testing due to the fact that there can only be one active scene during an integration test.

NoelStephensUnity committed:
test

Updating the IntegrationTestSceneHandler with the recent changes to ISceneManagerHandler.
Fixing some minor issues preventing some of the manual tests from functioning properly.
